allennlp/training/event_based_trainer/event_based_trainer.py

```diff
@@ -0,0 +1,87 @@
+"""
+blah
+"""
+from typing import Dict, Any, Union, List, Iterable, Callable
+from collections import defaultdict
+
+from allennlp.common.tqdm import Tqdm
+from allennlp.data.iterators.data_iterator import TensorDict
+from allennlp.models.model import Model
+from allennlp.training.event_based_trainer.callback import Callback
+from allennlp.training.event_based_trainer.events import Events
+from allennlp.training.trainer_base import TrainerBase
+
+
+class EventBasedTrainer(TrainerBase):
+    def __init__(self,
+                 model: Model,
+                 epochs_generator: Iterable[Iterable[TensorDict]],
+                 serialization_dir: str,
+                 cuda_device: Union[int, List] = -1) -> None:
+        super().__init__(serialization_dir, cuda_device)
+        self.model = self._move_to_gpu(model)
+        self.epochs_generator = epochs_generator
+        self.callbacks: Dict[str, List[Callback]] = defaultdict(list)
+
+        # Keeping track
+        self.batch_number = 0
+        self.epoch_number = 0
+        self.should_terminate = False
+        self.tqdm = None
+
+        # Current batch, current output
+        self.batch: TensorDict = None
+        self.output: Dict[str, Any] = None
+
+        # Any other state a callback might want to use
+        self.state: Dict[str, Any] = {}
+
+    def add_callback(self, event: str, callback: Callback) -> None:
+        self.callbacks[event].append(callback)
+
+    def remove_callback(self, event: str, callback: Callback) -> None:
+        event_callbacks = self.callbacks.get(event)
+        if event_callbacks:
+            self.callbacks[event] = [cb for cb in event_callbacks if cb != callback]
+
+    def fire_event(self, event: str) -> None:
+        for callback in self.callbacks.get(event, ()):
+            callback.call(self)
+
+    def train(self) -> Dict[str, Any]:
+        self.batch_number = 0
+        self.epoch_number = 0
+        train_metrics = []
+
+        self.fire_event(Events.TRAINING_START)
+        for epoch in self.epochs_generator:
+            self.fire_event(Events.EPOCH_START)
+
+            self.tqdm = Tqdm.tqdm(epoch)
+            for batch in self.tqdm:
+                # TODO(move to cuda device)
+
+                self.batch = batch
+                self.fire_event(Events.BATCH_START)
+                self.fire_event(Events.PROCESS_BATCH)
```

Comment:
one obvious alternative is to do it the way pytorch/ignite does it, which is that PROCESS_BATCH is not an event, instead you pass in something like a `Callable[[TensorDict], Any]` to the constructor

Reply:
Yeah, but if you go down that path, where do you stop?  And really, to do things `from_params`, you'd want that `Callable` to be `registered` somewhere, and so you add a new type for that callback, and things just get really messy (because how many callback types do you have, etc.).  I don't really like saving the state to `self`, but I haven't seen (and can't come up with) a better alternative.

Reply:
An alternative I see is having different functions, instead of dynamic events (which btw I don't see as a great advantage, but please let me know your opinion on this :smile:), like `on_epoch_start`, etc. In this case, every function can receive its own params (e.g., `batch` in `process_batch`).

This is something similar to what Keras does, but they define all functions under the same class (so, `Callback` instead of having `call`, would have: `on_training_start`, `on_epoch_start`, etc.).

I don't know how this would interact with the `from_params` thing, or how do you imagine the `from_params` should be with the implementation as it is now.

Reply:
right, the main reason for doing it this way is that you can do `Callback.from_params`, so that you could configure your whole trainer using config files, something like

```
trainer: {
    callbacks: {
        "TRAINING_START": [
            {"type": "set_up_tensorboard", ... }
        ],
        "EPOCH_END": [
            {"type": "check_validation_metrics", ... }
        ], ...
}
```

but you're right that the tradeoff that way is that it's hard to have a single interface that works for every kind of callback